validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetectionService.java

```diff
@@ -0,0 +1,181 @@
+                  .getValidatorIndices()
+                  .thenCompose(
+                      validatorIndices ->
+                          validatorApiChannel
+                              .checkValidatorsDoppelganger(
+                                  validatorIndices
+                                      .intStream()
+                                      .mapToObj(UInt64::valueOf)
+                                      .collect(Collectors.toList()),
+                                  currentEpoch)
+                              .thenApply(
+                                  validatorLivenessAtEpoches ->
+                                      doppelgangerDetected(
+                                          validatorLivenessAtEpoches,
+                                          validatorIndices
+                                              .intStream()
+                                              .mapToObj(UInt64::valueOf)
+                                              .collect(Collectors.toList())))
+                              .exceptionally(
+                                  throwable -> {
+                                    LOGGER.error(
+                                        "Unable to check validators doppelganger: {}",
+                                        throwable.getMessage());
+                                    return false;
+                                  })
+                              .thenApply(
+                                  doppelgangerDetected -> {
+                                    if (doppelgangerDetected) {
+                                      LOGGER.fatal(
+                                          "Doppelganger detected. Shutting down Validator Client.");
+                                      System.exit(1);
+                                    }
+                                    return null;
+                                  }));
+            })
+        .exceptionally(
+            throwable -> {
+              LOGGER.error("Unable to check validators doppelganger: {}", throwable.getMessage());
+              return null;
+            });
+  }
+
+  private boolean doppelgangerDetected(
+      final Optional<List<ValidatorLivenessAtEpoch>> validatorLivenessAtEpoches,
+      final List<UInt64> validatorIndices) {
+    if (validatorLivenessAtEpoches.isPresent()) {
+      List<ValidatorLivenessAtEpoch> doppelgangers =
+          validatorLivenessAtEpoches.get().stream()
+              .filter(
+                  validatorLivenessAtEpoch ->
+                      validatorIndices.contains(validatorLivenessAtEpoch.getIndex()))
+              .collect(Collectors.toList());
+      if (!doppelgangers.isEmpty()) {
+        LOGGER.fatal(
+            "Doppelganger detected. Validators indices: {}",
+            doppelgangers.stream()
+                .map(validatorLivenessAtEpoch -> validatorLivenessAtEpoch.getIndex().toString())
+                .collect(Collectors.joining(",")));
```

Comment:
given this is fatal, i guess we want to shutdown at this point - potentially what the user will find more useful is the actual abbreviated pubkeys rather than the validator indices (we should abbreviated pubkeys in the screen logger)

I think this probably needs to be in the screen logger too, not in the console logger... I think a reasonable example may be `StatusLogger.validatorsInitialized` - we'd call statusLogger and pass it a list of validators that were detected...

Reply:
Used the `StatusLogger` and tried to get the doppelgangers public keys (Handled the case where they could not be retrieved but normally that's not supposed to happen. If it's the case we only log the indices)